src/misc/degree_sequence.cpp

```diff
@@ -68,66 +70,232 @@ template<typename T> inline bool degree_less(const T &a, const T &b) {
+                    IGRAPH_CHECK(igraph_vector_int_push_back(&spokes, node.vertex)); // add as spoke
+
+                    node.degree--;
+                    insert_node(node);                // first, insert into bucket below
 
-                VECTOR(*edges)[2 * (ec + i)] = vd.vertex;
-                VECTOR(*edges)[2 * (ec + i) + 1] = vertices[i].vertex;
+                    buckets[bucket].count--;
+                    to_get--;
+                }
+                buckets[bucket].nodes.pop();                  // then pop from original bucket
             }
+            num_skip = 0;
         }
+        return IGRAPH_SUCCESS;
+    }
+};
 
-        ec += vd.degree;
+enum class Method {
+    LARGEST_FIRST,
+    SMALLEST_FIRST,
+    CUSTOM
+};
+
+static igraph_error_t igraph_i_havel_hakimi(const igraph_vector_int_t *degseq,
+                                            igraph_vector_int_t *edges,
+                                            Method method,
+                                            const igraph_vector_int_t *hub_order = nullptr) {
+    igraph_integer_t n_nodes = igraph_vector_int_size(degseq);
+
+    // ----- upfront error/graphicality checks ----- //
+    if (method == Method::CUSTOM && hub_order == nullptr) {
+        IGRAPH_ERROR("Custom ordering specified but no order given", IGRAPH_EINVAL);
+    } else if (n_nodes == 0 || (n_nodes == 1 && VECTOR(*degseq)[0] == 0)) {
+        return IGRAPH_SUCCESS;
     }
 
-    return IGRAPH_SUCCESS;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        igraph_integer_t deg = VECTOR(*degseq)[i];
+        if (deg >= n_nodes) {
+            IGRAPH_ERROR("Invalid degree sequence.", IGRAPH_EINVAL);
+        }
+    }
 
-fail:
-    IGRAPH_ERROR("The given degree sequence cannot be realized as a simple graph.", IGRAPH_EINVAL);
+    // ----- main Havel-Hakimi loop ----- //
+    // O(V + alpha(V) * E)
+    // O(V + E) for the LARGEST_FIRST method
+    igraph_vector_int_t seq;
+    IGRAPH_CHECK(igraph_vector_int_init_copy(&seq, degseq));
+
+    HavelHakimiList vault(seq);
+
+    igraph_integer_t n_edges_added = 0;
+    for (igraph_integer_t i = 0; i < n_nodes; i++) {
+        // hub node selection
+        vd_pair hub;
+        if (method == Method::SMALLEST_FIRST) {
+            if (!vault.get_min_node(/* out param */hub)) break;
+            vault.remove_min_node();
+        }
+        else if (method == Method::LARGEST_FIRST) {
+            if (!vault.get_max_node(/* out param */hub)) break;
+            vault.remove_max_node();
+        }
+        else if (method == Method::CUSTOM) {
+            igraph_integer_t index = VECTOR(*hub_order)[i];
+            igraph_integer_t degree = VECTOR(seq)[index];
+            hub = vd_pair{index, degree};
+            vault.buckets[degree].count--;
+        }
+        else {
+            IGRAPH_ERROR("Invalid method.", IGRAPH_EINVMODE);
+        }
```

Comment:
If this error triggers, that's a bug in the igraph code, so once again this should be an `IGRAPH_ASSERT` / `IGRAPH_FATAL` or skipped. The method is already validated in the top-level function.